pkg/proto/common.go

```diff
@@ -98,11 +103,15 @@ func BytesWithUInt16Len(buf []byte) ([]byte, error) {
 	return r, nil
 }
 
-// PutBytesWithUInt32Len prepends given buf with 4 bytes of it's length.
-func PutBytesWithUInt32Len(buf []byte, data []byte) {
-	sl := uint32(len(data))
-	binary.BigEndian.PutUint32(buf, sl)
+// PutBytesWithUInt32Len prepends given buf with 4 bytes of its length.
+func PutBytesWithUInt32Len(buf []byte, data []byte) error {
+	l := len(data)
+	if l > math.MaxInt32 {
+		return errors.Errorf("invalid data size %d", l)
+	}
+	binary.BigEndian.PutUint32(buf, uint32(l))
 	copy(buf[4:], data)
```

Comment:
I think `len(buf) >= 4 + len(data)` is necessary here.

Reply:
added